[FX] Generalize dict key check in `create-arg`

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#51927 [FX] Generalize dict key check in `create-arg`**

Previously, this check was very strict and didn't allow anything other than `str`. This change makes it so that we search through the key and make sure it doesn't contain `Proxy`, otherwise it's fine.

Differential Revision: [D26329655](https://our.internmc.facebook.com/intern/diff/D26329655)

Commit:
[FX] Generalize dict key check in `create-arg`

[ghstack-poisoned]

zdevito commented on torch/fx/proxy.py:
```diff
@@ -61,8 +61,17 @@ def create_arg(self, a: Any) -> Argument:
+                # Check for invalid dict keys. We do not want a Proxy to appear
+                # anywhere within the key. Since keys can be collection types,
+                # we iterate through the key with map_aggregate
+                is_invalid = [False] # 1-element array so it's mutable from the lambda
```
```
k = self.create_arg(k)
def no_node(arg):
  if isinstance(arg, Node):
    raise RuntimeError("Keys for dictionaries used as an argument cannot contain a "
                                        "Node. Got key: {k}")
map_aggregate(k, no_node)
```

Commit:
Update on "[FX] Generalize dict key check in `create-arg`"


Previously, this check was very strict and didn't allow anything other than `str`. This change makes it so that we search through the key and make sure it doesn't contain `Proxy`, otherwise it's fine.

Differential Revision: [D26329655](https://our.internmc.facebook.com/intern/diff/D26329655)

[ghstack-poisoned]

Commit:
Update on "[FX] Generalize dict key check in `create-arg`"


Previously, this check was very strict and didn't allow anything other than `str`. This change makes it so that we search through the key and make sure it doesn't contain `Proxy`, otherwise it's fine.

Differential Revision: [D26329655](https://our.internmc.facebook.com/intern/diff/D26329655)

[ghstack-poisoned]
